botocore/credentials.py

```diff
@@ -791,6 +820,206 @@ def _assume_role_base_kwargs(self, config):
+            # input, but for now we just error out when the temp creds expire.
+            refresh_func = lambda: exception_raiser(RefreshUnsupportedError)
+        else:
+            refresh_func = lambda: credential_normalizer(
+                self._create_client().assume_role_with_saml(
+                    **self._assume_role_base_kwargs(config))['Credentials'])
+        return RefreshableCredentials(
+            access_key=response['Credentials']['AccessKeyId'],
+            secret_key=response['Credentials']['SecretAccessKey'],
+            token=response['Credentials']['SessionToken'],
+            method=self.METHOD,
+            expiry_time=_parse_if_needed(
+                response['Credentials']['Expiration']),
+            refresh_using=refresh_func)
+
+    def _create_client(self):
+        # sts.assume_role_with_saml() requires no access keys,
+        # but we still need a pair of dummy values here to break recursion.
+        return self._client_creator(
+            'sts', aws_access_key_id='dummy', aws_secret_access_key='dummy')
+
+    def _retrieve_temp_credentials(self):
+        logger.debug("Retrieving credentials via AssumeRoleWithSaml.")
+        response = self._create_client().assume_role_with_saml(
+            **self._assume_role_base_kwargs(self._get_role_config_values()))
+        creds = self._create_creds_from_response(response)
+        return creds, response
+
+    def _assume_role_base_kwargs(self, config):
+        assertion = None
+        for authenticator in self.authenticators:
+            if authenticator.is_suitable(config):
+                assertion = authenticator.authenticate(
+                    config=config, username_prompter=self.username_prompter,
+                    password_prompter=self.password_prompter)
+                break
+        else:
+            raise ValueError("Unsupported saml_authentication_type: %s"
+                             % config.get('saml_authentication_type'))
+        if not assertion:
+            raise SamlError(detail='Login failed: SAML assertion not found')
+        idp_roles = self._parse_roles(assertion)
+        if not idp_roles:
+            raise SamlError(detail='Identity provider provides no role.')
+        role = self.role_selector(config.get('role_arn'), idp_roles)
+        if not role:
+            raise SamlError(detail='Unable to choose role "%s" from %s' % (
+                config.get('role_arn'), [r['RoleArn'] for r in idp_roles]))
+        role['SAMLAssertion'] = assertion
+        return role
+
+    def _parse_roles(self, assertion):
+        attribute = '{urn:oasis:names:tc:SAML:2.0:assertion}Attribute'
+        attr_value = '{urn:oasis:names:tc:SAML:2.0:assertion}AttributeValue'
+        awsroles = []
+        root = ET.fromstring(base64.b64decode(assertion))
+        for attr in root.getiterator(attribute):
+            if attr.get('Name')=='https://aws.amazon.com/SAML/Attributes/Role':
```

Comment:
spaces around the `==`.  I'd suggest just running flake8/pep8 through this PR code and fixing up any styling issues that get flagged.